frontend/public/components/hpa.tsx

```diff
@@ -31,8 +34,113 @@ const menuActions = [
   ...common,
 ];
 
-export const HorizontalPodAutoscalersDetails: React.SFC<HorizontalPodAutoscalersDetailsProps> = ({obj: hpa}) => {
-  return <div className="co-m-pane__body">
+const MetricsRow: React.SFC<MetricsRowProps> = ({type, current, target}) => <div className="row">
+  <div className="col-xs-6">
+    {type}
+  </div>
+  <div className="col-xs-3">
+    {current || '-'}
+  </div>
+  <div className="col-xs-3">
+    {target || '-'}
+  </div>
+</div>;
+
+const externalRow = (metric, current, key) => {
+  const { external } = metric;
+  const type = external.metricName;
+
+  let currentValue, targetValue;
+  // TODO: show metric selector for external metrics?
+  if (external.targetAverageValue) {
+    currentValue = _.get(current, 'object.currentAverageValue');
+    targetValue = external.targetAverageValue;
+  } else if (external.targetValue) {
+    currentValue = _.get(current, 'object.currentValue');
+    targetValue = external.targetValue;
+  }
+
+  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
+};
+
+const objectRow = (metric, current, key, ns) => {
+  const { object } = metric;
+  const type = <React.Fragment>
+    {object.metricName} on
+    <ResourceLink kind={object.target.kind} name={object.target.name} namespace={ns} title={object.target.name} />
+  </React.Fragment>;
+  const currentValue = _.get(current, 'object.currentValue');
+  const targetValue = object.targetValue;
+
+  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
+};
+
+const podRow = (metric, current, key) => {
+  const { pods } = metric;
+  const type = `${pods.metricName} on pods`;
+  const currentValue = _.get(current, 'pods.currentAverageValue');
+  const targetValue = pods.targetAverageValue;
+
+  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
+};
+
+const resourceRow = (metric, current, key) => {
```

Comment:
Wondering if this can be more declarative as well, if we split each bit out && use `const`:

```JavaScript 
const getUtilization = (current) => {
  let val = _.get(current, 'resource.currentAverageUtilization'); 
  return _.isNumber(val) ? `${val}%` : val;
}

const resourceRow = (metric, current, key) => {
  const currentValue = getUtilization(current);
  const targetValue = metric.resource.targetAverageUtilization
    ? `${metric.resource.targetAverageUtilization}%` 
    :  metric.resource.targetAverageValue;

  const type = metric.resource.targetAverageUtilization 
    ? <React.Fragment>resource {metric.resource.name}</React.Fragment> 
    : <React.Fragment>
        {type}&nbsp;<span className="small text-muted">(as a percentage of request)</span>
      </React.Fragment>;

  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
};
```
Not tested, very possible I missed something.

Reply:
I don't think this is parallel to the existing.

Reply:
Its very close, prob not perfect as I didn't run it to test.  The point is, its possible to isolate each variable & set it clearly.  Testing `metric.resource.targetAverageUtilization` twice to isolate the values makes it much easier to understand now && change in the future.

Reply:
Can inline the `getUtilization()` as well if preferred:

```JavaScript
const utilization = _.get(current, 'resource.currentAverageUtilization'); 
const currentValue = _.isNumber(utilization) ? `${utilization}%` : utilization;
```
Commenting more on isolating each variable than an exact impl.

Reply:
I inlined getUtilization and added currentAverageValue in parens as that's what the CLI does.

<img width="994" alt="screen shot 2018-06-01 at 10 47 13 am" src="https://user-images.githubusercontent.com/895728/41358272-06b4960a-6ef6-11e8-8418-d39627cbe308.PNG">

Reply:
That works, though if it *might* be reusable later, it's much better as a function.  I tend to not want to leave it for someone else to refactor later if they find that they need it.  That leads us to a lot of duplication.

I'm still pretty concerned about the complexity of the `if` statement.  There are many things happening in parallel that can be extracted.

Reply:
@spadgett, your thoughts?

Reply:
@rhamilto describe again quick the tricky part of this -- which vars may or may not be set together.  That might be a good clue.  I don't think its explicit in this current implementation, but maybe can be made explicit.

It sounded confusing, and I think my suggested refactor missed it, which probably confirms that there is some confusing logic masked here.

Reply:
Curious what the `_.isNumber(utilization)` really means.  There is something specific going on there, I think it would be helpful to name it (named fn);

The more we look at it and talk about it, the more I want a little more clarity on some of the bits going on.

Reply:
We're just checking if there's a value. 0 should evaluate to true even though it's falsey, so we use `_.isNumber`.

```
_.isNumber(undefined)
false
_.isNumber(null)
false
_.isNumber(0)
true
```

Reply:
If its a number, its assumed to be a `%`, otherwise `utililzation` will include its own suffix, indicating the kind of metric it is?

Reply:
In the end, I just don't think these need to be intertwined.  95% confident they can still be `const`, can pivot on `metric.resource.targetAverageUtilization` separately & be set cleanly w/o a ton of effort.  It would be more declarative, rely less on side effects & look more consistent w/our other react code / react code in general.  I'd rather set our trajectory away from intertwined blocks.

That said, @spadgett @rhamilto if you strongly disagree, feel free to move on.

Reply:
@benjaminapetersen worked together to come up with the following.  @spadgett, how about something like this?

```javascript
const getUtilization = (metric, current, metricUtilization) => {
  const currentAverageValue = _.get(current, 'resource.currentAverageValue');
  const utilization = _.get(current, 'resource.currentAverageUtilization');
  const currentValueUtilization = _.isNumber(utilization)
    ? `${utilization}% (${currentAverageValue})`
    : currentAverageValue;

  return metricUtilization
    ? currentValueUtilization
    : currentAverageValue;
}

const resourceRow = (metric, current, key) => {
  const metricUtilization = metric.resource.targetAverageUtilization;
  const currentValue = getUtilization(metric, current, metricUtilization);
  const targetValue = metricUtilization
    ? `${metricUtilization}%`
    :  metric.resource.targetAverageValue;
  const resource = `resource ${metric.resource.name}`;

  const type = metricUtilization
    ? <React.Fragment>{resource}</React.Fragment>
    : <React.Fragment>
        {resource}&nbsp;<span className="small text-muted">(as a percentage of request)</span>
      </React.Fragment>;

  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
};
```